### DISCUSSION TITLE
Consider upstreaming plasma layout

### DISCUSSION THREAD

**Discussion by nazar-pc** [qtile/qtile#4756] [Ideas] [answered]
I've been using https://github.com/numirias/qtile-plasma for years, it is quite awesome!

However I have difficulties upgrading with breaking changes introduced in Qtile since I need to patch my local version of qtile-plasma with more and move fixes.

Please consider upstreaming such an awesome layout, it is basically an advanced version of Tile and the only layout I use 99.9% of the time.

**Comment by tych0** [comment]
We'd be happy to accept a pull request with the code :)

**Reply by nazar-pc** [reply]
Okay, I'll try to create one when I'm in the right mood

**Comment by jrwrigh** [comment]
I made a similar request a few months ago: https://github.com/qtile/qtile/discussions/4487, but there was a pretty mixed reaction to it's inclusion in the main repository. Happy to see that @tych0 is interested in bringing it over, but I'm curious if @m-col or @elParaguayo have changed their thoughts.

I say this not to try stop this from happening, but just to make sure everyone's on-board with the idea before @nazar-pc (or whoever else) puts in the work to make a PR.

**Reply by nazar-pc** [reply]
I'm really not doing Python unless I have to, so if there is wider interest, someone else might be a better person doing it

**Reply by jrwrigh** [reply]
Yeah, I'm quite comfortable with Python, but not sure how much time I can afford to dedicate to getting up to speed on the code.

**Reply by nazar-pc** [reply]
The code works, but you'll have to merge most of currently opened PRs there for things to actually work with latest Qtile and Python versions

**Reply by jrwrigh** [reply]
Yeah, I know it works; I've also been running it for several years now. I've already got most (if not all) of the open PRs merged in on [my fork](https://github.com/jrwrigh/qtile-plasma) (sans your most recent ones).

**Reply by elParaguayo** [reply]
Let me take another look at the code. If it's not too complicated then I'll add it and write some tests for it (unless it's already got some)

**Reply by jrwrigh** [reply]
It does have some tests, though in 5 minutes of playing around with them I couldn't get it running. I'm not super familiar with pytest or the qtile testing framework (which it appears to be using).

**Reply by elParaguayo** [reply]
There's so much code to this one. Going to take some time...

**Reply by jrwrigh** [reply]
Yeah, note that the `enum.py` is just a copy/paste of the standard library's enum from 3.6. So that can be outright removed without much issue.

**Reply by elParaguayo** [reply]
I'm just working through the logic and changing some little bits here and there.

**Reply by elParaguayo** [reply]
@jrwrigh I'm getting a weird error and I'm curious if you see the same. Open `qtile shell`. (Assuming `Plasma` is the active layout) do `cd layout` and then `mode_vertical()`. Try opening a new window and see if you get an error message appear in the `qtile shell` session.

The code seems to work but I've no idea why I get an error message. Sorry, can't paste the message at the moment as I'm at work!

**Reply by jrwrigh** [reply]
TBH, I don't use `qtile shell` ever (maybe I've been using qtile wrong for years... haha).

But doing I think what you asked:
```
❯ qtile shell
 > cd layout
layout
layout > mode_vertical()
layout >
```

And then opening a new window (not within `qtile shell`), I do not get an error in the shell or in the log. Note I'm still on qtile `0.24.0` and am on my own fork which incorporates several fixes (not sure if you're going off of that or the main repo right now).

**Reply by elParaguayo** [reply]
OK. Thanks for checking.

**Reply by jrwrigh** [reply]
Curious, what was the error message?

Edit: To clarify above, I'm on my own fork _of Plasma_, not qtile itself.

**Reply by elParaguayo** [reply]
It's unrelated to the layout, it's a side effect of an earlier commit (@tych0 - I'll post in your PR about this)

**Comment by elParaguayo** [comment]
> It does have some tests, though in 5 minutes of playing around with them I couldn't get it running. I'm not super familiar with pytest or the qtile testing framework (which it appears to be using).

Wow. There are a lot of tests here. Really helpful for learning what the code does. Slowly working my way through it still.

**Reply by elParaguayo** [reply]
I may be misunderstanding the code but I'm not convinced by the some of the logic in this.

**Reply by jrwrigh** [reply]
If you have any specific questions, I'd be happy to look at the code and see if I can make any sense out of it. I'm guessing you haven't really used the layout before, so maybe a regular user's perspective might help.

**Reply by elParaguayo** [reply]
I'm trying to get my head round the interaction of `add_child` and `fit_into`. It seems to resize `fixed` width nodes when it shouldn't.

**Reply by jrwrigh** [reply]
You mean if it tries to add a `fixed` node as a child? ie.
```python
assert(child.fixed)
parent.add_child(child)
```

**Reply by jrwrigh** [reply]
I think that is true to my experience. If I have a fixed width child node and move it to another group (which iiuc would call `add_child` on the new group's parent node), the fixed width is not respected in the new group. 

I don't have much of an issue with that behavior. Most of the time if I have a fixed width window, it's because I've changed the size of a split from it's equispaced point. If I'm sending that to another group, I don't really care about the size in the new group.

**Reply by elParaguayo** [reply]
Consider this example:
```
>>> from libqtile.layout.plasma import Node
>>> Node.min_siize_default = 10
>>> root = Node("root", 0, 0, 100, 100)
>>> a = Node("A")
>>> b = Node("B")
>>> c = Node("C")
>>> root.add_node(a)
>>> root.add_node(b)
>>> print(f"{a.width=} {b.width=}")
a.width=50.0 b.width=50.0
>>> a.width = 60
>>> print(f"{a.width=} {b.width=}")
a.width=60 b.width=40.0
>>> print(f"{a.fixed=} {b.fixed=}")
a.fixed=True b.fixed=False
>>> root.add_node(c)
>>> print(f"{a.width=} {b.width=} {c.width=}")
a.width=56.66666666666666 b.width=21.66666666666667 c.width=21.66666666666667
```
I would still expect node `A` to have a width of 60 with `B` and `C` being 20 each. There's no reason for `A` to have been shrunk.

**Reply by elParaguayo** [reply]
Think I've fixed it.

EDIT: Hmm. I suppose it depends on what you think the appropriate behaviour is:
- Assume you've got a root that is 120 wide and a node with a minimum width of 10
- Add 2 nodes and they'll both be 60 wide but flexible
- Set the first to 80 and the second to 40. They're now both fixed with.
- Add a new node. QUESTION: how wide should that node be? There are two options
  - 40 (i.e. container width / 3) - but this makes a big change to the existing nodes which had a fixed size
  - 10 (i.e. the minimum width) - we have two fixed width nodes already so we should make as small a change as necessary

The layout seems to do the first but I think we should do the second.

**Reply by jrwrigh** [reply]
I'm fine with either tbh, but the 2nd option makes the most sense, at least in a vacuum. These are edge cases that I don't run 
 into very often.

**Comment by elParaguayo** [comment]
Draft PR: #4760

**Reply by elParaguayo** [reply]
This has now been merged.

**Comment by elParaguayo** [comment, marked as answer]
Closing as the layout has now been merged.
